src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,140 @@
+        if (first == null) {
+            first = last = newNode;
+        } else {
+            last.next = newNode;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Objects.checkIndex(index, size + 1);
+        Node<T> newNode = new Node<>(value);
+        if (first == null) {
+            first = last = newNode;
+        } else if (index == 0) {
+            newNode.next = first;
+            first = newNode;
+        } else if (index == size) {
+            last.next = newNode;
+            last = newNode;
+        } else {
+            Node<T> current = getNodeByIndex(index - 1);
+            newNode.next = current.next;
+            current.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> newNode = getNodeByIndex(index);
+        T oldValue = newNode.element;
+        newNode.element = value;
+        return oldValue;
+
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        T removedElement;
+        if (index == 0) {
+            removedElement = first.element;
+            first = first.next;
+            if (first == null) {
+                last = null;
+            }
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            removedElement = prev.next.element;
+            prev.next = prev.next.next;
+            if (index == size - 1) {
+                last = prev;
+            }
+        }
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        int index = 0;
+        while (current != null) {
+            if (current.element == object || current.element != null
+                    && current.element.equals(object)) {
+                remove(index);
+                return true;
+            }
+            current = current.next;
+            index++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return first == null;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Objects.checkIndex(index, size);
+        if (index == size - 1) {
+            return last;
+        } else {
+            return nodeAt(index);
+        }
+    }
+
+    private Node<T> nodeAt(int index) {
+        Node<T> currentNode = first;
```

Comment:
Ось що я мав на увазі: 
```suggestion
         Node<T> currentNode;
        if (index < size / 2) {
            x = currentNode;
            for (int i = 0; i < index; i++) {
                currentNode = currentNode.next;
            }
        } else {
            currentNode = last;
            for (int i = size - 1; i > index; i--) {
                currentNode = currentNode.prev;
            }
        }
        return currentNode;
```